### DISCUSSION TITLE
Waveshare 2.9" E-Paper Display (It's driving me mad plz)

### DISCUSSION THREAD

**Discussion by minizom** [micropython/micropython#12180] [Hardware & Peripherals]
I will pre-empt this with I am a fool, and there is probably a simple solution I'm not seeing. I am using the Waveshare Waveshare 2.9" E-Paper Display Module for Raspberry Pi Pico (Black/White) (296×128).  I have tested the hardware thoroughly and all of it works ok. It displays the text horizontally when the display is verticle. I simply want to have the text display horizontally (across) the e-ink display when it is landscape. I am new wot micro python (yes I'm one of those people)

I am aware there are a number of similar issues/threads/guides with other waveshape displays such as
https://forum.micropython.org/viewtopic.php?f=18&t=6319&sid=b987deb8bb64da2bb21b5db6f5433764 
or  https://github.com/waveshareteam/e-Paper/blob/master/RaspberryPi_JetsonNano/python/lib/waveshare_epd/epd2in9d.py

But I simply cannot get it to work and I'm on the verge of giving up and purchasing a shitty LCD to do the job. 
the guide they provide is: https://www.waveshare.com/wiki/2.9inch_e-Paper_Module_Manual#Working_With_Raspberry_Pi
and I have attached ta trimmed-down example code. 

I have tried a number of methods, but most were rubbish.
I'm guessing the answer is in one of several lines. 
swapping framebuf.MONO_HLSB) to framebuf.MONO_VLSB) (might have the wrong idea about what that does ) 
and starts with the correct landscape code 'self.send_data(0x07)' in 'def init' 

Any help would be appreciated, or If u just want to mock that is also ok. let me know if I can provide any other information. 

Code below: 

`from machine import Pin, SPI
import framebuf
import utime

EPD_2IN9D_lut_vcomDC =[
    0x00, 0x08, 0x00, 0x00, 0x00, 0x02,
    0x60, 0x28, 0x28, 0x00, 0x00, 0x01,
    0x00, 0x14, 0x00, 0x00, 0x00, 0x01,
    0x00, 0x12, 0x12, 0x00, 0x00, 0x01,
    0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
    0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
    0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
    0x00, 0x00,
]
EPD_2IN9D_lut_ww =[
    0x40, 0x08, 0x00, 0x00, 0x00, 0x02,
    0x90, 0x28, 0x28, 0x00, 0x00, 0x01,
    0x40, 0x14, 0x00, 0x00, 0x00, 0x01,
    0xA0, 0x12, 0x12, 0x00, 0x00, 0x01,
    0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
    0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
    0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
]
EPD_2IN9D_lut_bw =[
    0x40, 0x17, 0x00, 0x00, 0x00, 0x02,
    0x90, 0x0F, 0x0F, 0x00, 0x00, 0x03,
    0x40, 0x0A, 0x01, 0x00, 0x00, 0x01,
    0xA0, 0x0E, 0x0E, 0x00, 0x00, 0x02,
    0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
    0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
    0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
]
EPD_2IN9D_lut_wb =[
     0x80, 0x08, 0x00, 0x00, 0x00, 0x02,
    0x90, 0x28, 0x28, 0x00, 0x00, 0x01,
    0x80, 0x14, 0x00, 0x00, 0x00, 0x01,
    0x50, 0x12, 0x12, 0x00, 0x00, 0x01,
    0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
    0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
    0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
]
EPD_2IN9D_lut_bb =[
    0x80, 0x08, 0x00, 0x00, 0x00, 0x02,
    0x90, 0x28, 0x28, 0x00, 0x00, 0x01,
    0x80, 0x14, 0x00, 0x00, 0x00, 0x01,
    0x50, 0x12, 0x12, 0x00, 0x00, 0x01,
    0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
    0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
    0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
]
'''

'''
EPD_2IN9D_lut_vcom1 =[
    0x00, 0x19, 0x01, 0x00, 0x00, 0x01,
    0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
    0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
    0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
    0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
    0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
    0x00, 0x00, 0x00, 0x00, 0x00, 0x00
    ,0x00, 0x00,
]
EPD_2IN9D_lut_ww1 =[
    0x00, 0x19, 0x01, 0x00, 0x00, 0x01,
    0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
    0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
    0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
    0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
    0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
    0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
]
EPD_2IN9D_lut_bw1 =[
    0x80, 0x19, 0x01, 0x00, 0x00, 0x01,
    0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
    0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
    0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
    0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
    0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
    0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
]
EPD_2IN9D_lut_wb1 =[
    0x40, 0x19, 0x01, 0x00, 0x00, 0x01,
    0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
    0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
    0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
    0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
    0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
    0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
]
EPD_2IN9D_lut_bb1 =[
    0x00, 0x19, 0x01, 0x00, 0x00, 0x01,
    0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
    0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
    0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
    0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
    0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
    0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
]

EPD_WIDTH       = 128
EPD_HEIGHT      = 296
RST_PIN         = 12
DC_PIN          = 8
CS_PIN          = 9
BUSY_PIN        = 13
FULL_UPDATE = 0
PART_UPDATE = 1

class EPD_2IN9_D(framebuf.FrameBuffer):
    def __init__(self):
        self.reset_pin = Pin(RST_PIN, Pin.OUT)
        
        self.busy_pin = Pin(BUSY_PIN, Pin.IN, Pin.PULL_UP)
        self.cs_pin = Pin(CS_PIN, Pin.OUT)
        self.width = EPD_WIDTH
        self.height = EPD_HEIGHT
        
        self.lut_vcomDC = EPD_2IN9D_lut_vcomDC
        self.lut_ww = EPD_2IN9D_lut_ww
        self.lut_bw = EPD_2IN9D_lut_bw
        self.lut_wb = EPD_2IN9D_lut_wb
        self.lut_bb = EPD_2IN9D_lut_bb
        
        self.lut_vcom1 = EPD_2IN9D_lut_vcom1
        self.lut_ww1 = EPD_2IN9D_lut_ww1
        self.lut_bw1 = EPD_2IN9D_lut_bw1
        self.lut_wb1 = EPD_2IN9D_lut_wb1
        self.lut_bb1 = EPD_2IN9D_lut_bb1
        
        self.spi = SPI(1)
        self.spi.init(baudrate=4000_000)
        self.dc_pin = Pin(DC_PIN, Pin.OUT)
        
        
        
        self.buffer = bytearray(self.width * self.height // 8)
        
        super().__init__(self.buffer, self.width, self.height, framebuf.MONO_VLSB)
        self.init()

    def digital_write(self, pin, value):
        pin.value(value)

    def digital_read(self, pin):
        return pin.value()

    def delay_ms(self, delaytime):
        utime.sleep_ms(delaytime)

    def spi_writebyte(self, data):
        self.spi.write(bytearray(data))

    def module_exit(self):
        self.digital_write(self.reset_pin, 0)

    # Hardware reset
    def reset(self):
        self.digital_write(self.reset_pin, 1)
        self.delay_ms(200)
        self.digital_write(self.reset_pin, 0)
        self.delay_ms(2)
        self.digital_write(self.reset_pin, 1)
        self.delay_ms(200)   

    def send_command(self, command):
        self.digital_write(self.dc_pin, 0)
        self.digital_write(self.cs_pin, 0)
        self.spi_writebyte([command])
        self.digital_write(self.cs_pin, 1)

    def send_data(self, data):
        self.digital_write(self.dc_pin, 1)
        self.digital_write(self.cs_pin, 0)
        self.spi_writebyte([data])
        self.digital_write(self.cs_pin, 1)
        
    def send_data1(self, buf):
        self.digital_write(self.dc_pin, 1)
        self.digital_write(self.cs_pin, 0)
        self.spi.write(bytearray(buf))
        self.digital_write(self.cs_pin, 1)
        
    def ReadBusy(self):
        print('e-Paper busy')
        busy = 1
        self.send_command(0x71)
        while(self.digital_read(self.busy_pin) == 0): 
            self.send_command(0x71)
        self.delay_ms(200)    
        print('e-Paper busy release')
    
    def SetFullReg(self):
        self.send_command(0x50)
        self.send_data(0xB7)
        
        self.send_command(0x20)
        self.send_data1(self.lut_vcomDC[0:44])
            
        self.send_command(0x21)
        self.send_data1(self.lut_ww[0:42])
        
        self.send_command(0x22)
        self.send_data1(self.lut_bw[0:42])

        self.send_command(0x23)
        self.send_data1(self.lut_bb[0:42])
            
        self.send_command(0x24)
        self.send_data1(self.lut_wb[0:42])
          
    
    def SetPartReg(self):
        self.send_command(0x82)
        self.send_data(0x00)
        self.send_command(0x50)
        self.send_data(0xB7)
        
        self.send_command(0x20)
        self.send_data1(self.lut_vcom1[0:44])
            
        self.send_command(0x21)
        self.send_data1(self.lut_ww1[0:42])
            
        self.send_command(0x22)
        self.send_data1(self.lut_bw1[0:42])
            
        self.send_command(0x23)
        self.send_data1(self.lut_wb1[0:42])
            
        self.send_command(0x24)
        self.send_data1(self.lut_bb1[0:42])
            
    
    def TurnOnDisplay(self):
        self.send_command(0x12)
        self.delay_ms(100)       
        self.ReadBusy()

    def init(self): 
        print('init')
        self.reset()

        self.send_command(0x01)
        self.send_data(0x07) # important line, dictates horizontal veres verticle. 
        self.send_data(0x00)
        self.send_data(0x2B)
        self.send_data(0x2B)
        self.send_data(0x03)
        
        
        self.send_command(0x06)
        self.send_data(0x17)
        self.send_data(0x17)
        self.send_data(0x17)

        self.send_command(0x04) 
        self.ReadBusy()
        
        self.send_command(0x00)
        self.send_data(0xBF)
        self.send_data(0x0E)
        
        self.send_command(0x30)
        self.send_data(0x3A)
        
        self.send_command(0x61)
        self.send_data(self.width)
        self.send_data((self.height&0x100)>>8)
        self.send_data(self.height&0xff)
        
        self.send_command(0x82)
        self.send_data(0x28)
        
        
    def display(self, image):
        high = self.height
        if( self.width % 8 == 0) :
            wide =  self.width // 8
        else :
            wide =  self.width // 8 + 1
        self.send_command(0x10)
        self.send_data1([0x00] * high * wide)
        
        self.send_command(0x13)
        self.send_data1(image)
            
        self.SetFullReg()
        self.TurnOnDisplay()
        
    def display_Partial(self, image):
        self.SetPartReg()
        self.send_command(0x91)
        self.send_command(0x90)
        self.send_data(0)
        self.send_data(self.width - 1)
        
        self.send_data(0)
        self.send_data(0)
        self.send_data(self.height//256)
        self.send_data(self.height%256 - 1)
        self.send_data(0X28)
        
        high = self.height
        if( self.width % 8 == 0) :
            wide =  self.width // 8
        else :
            wide =  self.width // 8 + 1
        self.send_command(0x13)
        self.send_data1(image)
        
        self.TurnOnDisplay()

    
    def Clear(self, color):
        high = self.height
        if( self.width % 8 == 0) :
            wide =  self.width // 8
        else :
            wide =  self.width // 8 + 1
        self.send_command(0x10)
        self.send_data1([color] * high * wide)
        
        self.send_command(0x13)
        self.send_data1([~color] * high * wide)
            
        self.SetFullReg()
        self.TurnOnDisplay()

    def sleep(self):
        self.send_command(0x50)
        self.send_data(0xF7)
        self.send_command(0x02)
        self.send_command(0x07)
        self.send_data(0xA5)

        
if __name__=='__main__':
    epd = EPD_2IN9_D()
    epd.Clear(0x00)
    
    epd.fill(0xff)
    epd.text("hello", 5, 50, 0x00)
    epd.display(epd.buffer)
    epd.delay_ms(2000) 

    epd.Clear(0x00)
    epd.delay_ms(2000)
    epd.sleep()`

**Comment by kjm1102** [comment]
I haven't tangled with that particular display but I've found waveshare fairly responsive to technical questions if your patient.

**Reply by minizom** [reply]
thnx for the suggestion :) responsive up to a point. They replied quickly but directed me to the existing product guide that appears to not work. I'm already waiting on a reply from them before posting on here.

**Comment by peterhinch** [comment]
Modifying display divers requires a fair degree of experience. Unless you're confident to tackle this, my advice is to find out what displays are already supported and choose one of those.

[This link](https://github.com/peterhinch/micropython-nano-gui/blob/master/DISPLAYS.md) details some supported displays including eInk/ePaper. You could also look at [awesome micropython](https://awesome-micropython.com/).

**Reply by minizom** [reply]
thnx for the response :) sadly that's what I'm learning. The e ink I purchased was advertised with the ability to display text either way I just assumed the correct display drivers for both portrait and landscape would have been provided not just portrait. You might be right, and unless waveshare get back to me or there are any other replies to this I may do just that, but sadly I've already created the housing components and board design with this display in mind. Thnx for the links I'll check both of them out now :)

**Reply by peterhinch** [reply]
Your display seems similar to [this one](https://www.adafruit.com/product/4777) which is supported. It's possible that [this driver](https://github.com/peterhinch/micropython-nano-gui/blob/master/DRIVERS.md#51-adafruit-monochrome-eink-displays) will work with your display, or failing that, perhaps the Adafruit display will fit your housing. That driver supports landscape or portrait mode.
